### ISSUE TITLE
Comparison of KINC-python / KINCv1 / KINCv3 output data

### ISSUE THREAD

**Issue by bentsherman** [SystemsGenetics/KINC#34]
A few weeks ago I generated cluster data for Yeast using KINCv1, and I used the Import Correlation Matrix analytic to convert the cluster data into CCM and CMX files.

The KINCv1 CCM is 3.4 GB while the KINCv3 CCM is 8 GB, which is a pretty big discrepancy. Now, considering that we're using a different GMM implementation in KINCv3, we should expect some difference. But it seems like the KINCv1 CCM is just generally less dense, so either KINCv1 finds multiple clusters less often or it is filtering out clusters at the end, I will be looking into this.

In the meantime, we can still do a data evaluation for the paper by comparing KINCv1 data with it's equivalent imported data. And that comparison is still good: the EMX is smaller than the raw text, and the combined CCM / CMX is smaller than the KINCv1 cluster data.

Here is a more exhaustive listing of what current data sizes look like for Yeast. Note that I don't have a reliable data point for KINCv3 CMX yet.
```
6.7G	data/Yeast/clusters-sc
8.1G	data/Yeast/Yeast.ccm
6.7G	data/Yeast/Yeast-clusters.txt
20M	data/Yeast/Yeast-ematrix.txt
5.3M	data/Yeast/Yeast.emx
3.4G	data/Yeast/Yeast-imported.ccm
395M	data/Yeast/Yeast-imported.cmx
```

**Comment by spficklin** [comment]
I beleive the  default for KINC v1 was not saving results below a specific correlation threshold.  There is an argument where you can set the threshold to zero and it should save all of the output.

UPDATE: it is the --min_sim argument but it appears to be set to zero by default....  Can you verify that you have results for the lower correlation values?

**Comment by bentsherman** [comment]
Yes my imported CMX has all of the correlations, I didn't apply a threshold during similarity.

In it's current state we can't apply a correlation threshold to the CCM until we know the corresponding correlations, so we have to have generated the full CCM already. So we would need a way to perform both analytics at once if we want to minimize disk usage.

In fact, even now the correlation threshold is problematic because it could cause the CCM and CMX to get out of sync: suppose a gene pair has 5 clusters, but only two clusters have a correlation above 0.5, then the CMX will save two correlations and they will be mismatched with the clusters in the CCM.

So we need to resolve this issue before we can produce reliable data.

**Comment by bentsherman** [comment]
Currently developing Python scripts to mirror KINC similarity and RMT, they are easier to develop and use so they should give us a good second opinion about the validity of KINCv3 results.

**Comment by bentsherman** [comment]
Python scripts have now been developed for KINC similarity, threshold, and extract so we can compare results throughout the entire KINC pipeline without using KINCv1 (assuming that the python scripts work properly, which I will keep checking on as I run tests). So now I can easily test results from KINCv3 on small data.

I'm currently using Yeast-100 (first 100 genes from Yeast GEM), and I'm first testing Similarity with no clustering and Pearson correlation. In comparing KINCv3 and KINC-python, the sample strings are the same but the correlations differ in many cases. So I'll check the KINCv3 source code for bugs. I'll continue testing in this manner with Spearman, then GMM+Pearson and GMM+Spearman.

(As a side note, now that we have these python scripts we can easily test other clustering methods for efficacy so I think I can safely remove k-means from KINCv3.)

**Comment by bentsherman** [comment]
Pearson and Spearman have been validated (for serial and OpenCL). Just had to fix one thing in Pearson. Moving on to testing GMM.

**Comment by bentsherman** [comment]
In order to verify the GMM analytic, I think the important things to compare are (1) the number of clusters for each gene pair and (2) the characteristics of each cluster for each gene pair. In comparing results for Yeast-100, I found that the number of clusters for each gene pair was the same 64% of the time, and differed no more than 1 90% of the time. This level of similarity is roughly the same as when I compared GMM serial / OpenCL, and given the inherent variability of GMM results, I think it's fair to say that GMM is verified. As for comparing the cluster characteristics (namely the sample size and sample string of each cluster)... I don't think that's really warranted here.

At this point I feel pretty good about the validity of KINCv3 data, at least on the small scale. The Python code won't help us much for verifying large-scale data, but the script `visualize.py` especially will help us inspect larger data, so at least now we're better equipped to keep an eye on our results from time to time. Closing this issue.
